src/hardware/signal_generator.rs

```diff
@@ -0,0 +1,176 @@
+                    i32::MAX as u32,
+                ]
+            } else {
+                [
+                    (config.frequency * conversion_factor / config.symmetry)
+                        as u32,
+                    (config.frequency * conversion_factor
+                        / (1.0 - config.symmetry)) as u32,
+                ]
+            }
+        };
+
+        Ok(Config {
+            amplitude: DacCode::try_from(config.amplitude)
+                .or(Err(Error::InvalidAmplitude))?
+                .into(),
+            signal: config.signal,
+            frequency_tuning_word,
+        })
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Config {
+    /// The type of signal being generated
+    pub signal: Signal,
+
+    /// The full-scale output code of the signal
+    pub amplitude: i16,
+
+    /// The frequency tuning word of the signal. Phase is incremented by this amount
+    pub frequency_tuning_word: [u32; 2],
+}
+
+#[derive(Debug)]
+pub struct SignalGenerator {
+    phase_accumulator: u32,
+    config: Config,
+}
+
+impl Default for SignalGenerator {
+    fn default() -> Self {
+        Self {
+            config: BasicConfig::default().try_into().unwrap(),
+            phase_accumulator: 0,
+        }
+    }
+}
+
+impl SignalGenerator {
+    /// Construct a new signal generator with some specific config.
+    ///
+    /// # Args
+    /// * `config` - The config to use for generating signals.
+    ///
+    /// # Returns
+    /// The generator
+    pub fn new(config: Config) -> Self {
+        Self {
+            config,
+            phase_accumulator: 0,
+        }
+    }
+
+    /// Update waveform generation settings.
+    pub fn update_waveform(&mut self, new_config: Config) {
+        self.config = new_config;
+    }
+}
+
+impl core::iter::Iterator for SignalGenerator {
+    type Item = i16;
+
+    /// Get the next value in the generator sequence.
+    fn next(&mut self) -> Option<i16> {
+        self.phase_accumulator = self.phase_accumulator.wrapping_add(
+            if (self.phase_accumulator as i32).is_negative() {
+                self.config.frequency_tuning_word[0]
+            } else {
+                self.config.frequency_tuning_word[1]
+            },
+        );
+
+        let phase = self.phase_accumulator as i32;
+
+        let amplitude: i16 = match self.config.signal {
+            Signal::Cosine => (dsp::cossin(phase).0 >> 16) as i16,
+            Signal::Square => {
+                if phase.is_negative() {
+                    i16::MAX
+                } else {
+                    i16::MIN
+                }
+            }
+            Signal::Triangle => i16::MAX - (phase.abs() >> 15) as i16,
+        };
+
+        // Calculate the final output result as an i16.
+        let result = amplitude as i32 * self.config.amplitude as i32;
+
+        // Note: We downshift by 15-bits to preserve only one of the sign bits.
+        Some((result >> 15) as i16)
```

Comment:
We want to add rounding bias here.

```suggestion
        Some(((result + (1 << 14)) >> 15) as i16)
```